### ISSUE TITLE
Adapt to FreeBSD

### ISSUE THREAD

**Issue by penguinredish** [torognes/vsearch#361]
I tried to compile on FreeBSD without success.

Output for `autogen.sh`:

```
configure.ac:101: installing './compile'
configure.ac:6: installing './config.guess'
configure.ac:6: installing './config.sub'
configure.ac:7: installing './install-sh'
configure.ac:7: installing './missing'
src/Makefile.am: installing './depcomp'
```

Output for `configure`:

```
checking build system type... amd64-unknown-freebsd12.0
checking host system type... amd64-unknown-freebsd12.0
checking target system type... amd64-unknown-freebsd12.0
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/local/bin/gmkdir -p
checking for gawk... no
checking for mawk... no
checking for nawk... nawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking whether make supports the include directive... yes (GNU style)
checking dependency style of g++... none
checking for ranlib... ranlib
checking for pthread_create in -lpthread... yes
checking for dlopen in -ldl... yes
checking for GetProcessMemoryInfo in -lpsapi... no
checking how to run the C++ preprocessor... g++ -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking getopt.h usability... yes
checking getopt.h presence... yes
checking for getopt.h... yes
checking fcntl.h usability... yes
checking fcntl.h presence... yes
checking for fcntl.h... yes
checking float.h usability... yes
checking float.h presence... yes
checking for float.h... yes
checking regex.h usability... yes
checking regex.h presence... yes
checking for regex.h... yes
checking ctype.h usability... yes
checking ctype.h presence... yes
checking for ctype.h... yes
checking locale.h usability... yes
checking locale.h presence... yes
checking for locale.h... yes
checking limits.h usability... yes
checking limits.h presence... yes
checking for limits.h... yes
checking for string.h... (cached) yes
checking sys/time.h usability... yes
checking sys/time.h presence... yes
checking for sys/time.h... yes
checking dlfcn.h usability... yes
checking dlfcn.h presence... yes
checking for dlfcn.h... yes
checking for inline... inline
checking for size_t... yes
checking for uint32_t... yes
checking for int64_t... yes
checking for uint64_t... yes
checking for uint8_t... yes
checking for stdlib.h... (cached) yes
checking for GNU libc compatible malloc... yes
checking for working strtod... yes
checking for working alloca.h... no
checking for alloca... yes
checking for stdlib.h... (cached) yes
checking for GNU libc compatible realloc... yes
checking for memmove... yes
checking for memcpy... yes
checking for posix_memalign... yes
checking for gettimeofday... yes
checking for localtime_r... yes
checking for memchr... yes
checking for memset... yes
checking for pow... yes
checking for regcomp... yes
checking for strcasecmp... yes
checking for strchr... yes
checking for strcspn... yes
checking for sysinfo... no
checking for library containing pthread_create... none required
checking for ps2pdf... yes
checking pthread.h usability... yes
checking pthread.h presence... yes
checking for pthread.h... yes
checking bzlib.h usability... yes
checking bzlib.h presence... yes
checking for bzlib.h... yes
checking zlib.h usability... yes
checking zlib.h presence... yes
checking for zlib.h... yes
checking windows.h usability... no
checking windows.h presence... no
checking for windows.h... no
checking for gcc... gcc
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... none
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating src/Makefile
config.status: creating man/Makefile
config.status: creating config.h
config.status: executing depfiles commands
```

And finally the error during `make`:

```
make  all-recursive
Making all in src
make[2]: "/usr/home/felselva/biome/bin/vsearch-master/src/Makefile" line 1081: Variable/Value missing from "export"
make[2]: Fatal errors encountered -- cannot continue
make[2]: stopped in /usr/home/felselva/biome/bin/vsearch-master/src
*** Error code 1

Stop.
make[1]: stopped in /usr/home/felselva/biome/bin/vsearch-master
*** Error code 1

Stop.
make: stopped in /usr/home/felselva/biome/bin/vsearch-master
```

On line 1081 of `src/Makefile`:

```
export MACOSX_DEPLOYMENT_TARGET
```

Unfortunately I don't have any skills with autogen to solve the problem by myself.

**Comment by penguinredish** [comment]
I will try using the Linux compatibility layer from FreeBSD.

**Comment by torognes** [comment]
I'll see if I can solve this issue, so I am reopening.

**Comment by torognes** [comment]
You could try to change line 15 of `src/Makefile.am` from

`export MACOSX_DEPLOYMENT_TARGET`

to

`export MACOSX_DEPLOYMENT_TARGET="10.7"`

That should fix the error in the message you received.

I'll include that change in the next release of VSEARCH.

Unfortunately I do not have access to a FreeBSD system so I am unable to compile or test it there.

**Comment by torognes** [comment]
Perhaps fixed with commit da413974afa6985a15bf7ebb865d9787e412f849.

**Comment by torognes** [comment]
Included in version 2.12.0 just released.

**Comment by frederic-mahe** [comment]
@penguinredish could you please confirm compilation goes well?

**Comment by outpaddling** [comment]
FYI, there's a work-in-progress FreeBSD port here:

[https://github.com/outpaddling/freebsd-ports-wip](https://github.com/outpaddling/freebsd-ports-wip)

For other BSDs (and Linux, Mac OS or any other POSIX platform), there's also a package in pkgsrc-wip.

**Comment by torognes** [comment]
Thanks for the info, @outpaddling. I'll try to adapt the vsearch source code better to FreeBSD in the future by incorporating the changes made there.

**Comment by outpaddling** [comment]
I have a couple patches in development  to support FreeBSD and NetBSD. I'll post them here shortly, after I run through the QA process for the port and package.

**Comment by outpaddling** [comment]
Here are the final patches.

I tested these on CentOS, FreeBSD, NetBSD, and OS X.

It looks like somebody made a good effort to eliminate redundant #includes from the if-else structure in the original vsearch.h, but I think this would be too confusing with the added platforms, so I flattened it.  I made sure all the components were preserved for the __APPLE__  and __linux__ cases (and tested them as mentioned above).

I wonder if it would be better to #include vsearch.h in city.cc rather than redefine things.

```
--- src/city.cc.orig	2019-04-30 11:57:32 UTC
+++ src/city.cc
@@ -60,6 +60,13 @@ static uint32 UNALIGNED_LOAD32(const char *p) {
 #define bswap_32(x) OSSwapInt32(x)
 #define bswap_64(x) OSSwapInt64(x)
 
+#elif defined(__FreeBSD__)
+
+#include <sys/endian.h>
+
+#define bswap_32(x) bswap32(x)
+#define bswap_64(x) bswap64(x)
+
 #elif defined(__NetBSD__)
 
 #include <sys/types.h>
```

```
--- src/vsearch.h.orig	2019-04-30 11:57:32.000000000 +0000
+++ src/vsearch.h
@@ -138,30 +138,50 @@
 #define bswap_32(x) _byteswap_ulong(x)
 #define bswap_64(x) _byteswap_uint64(x)
 
-#else
-
-#ifdef __APPLE__
+#elif defined(__APPLE__)
 
 #define PROG_OS "macos"
 #include <sys/sysctl.h>
 #include <libkern/OSByteOrder.h>
+#include <sys/resource.h>
 #define bswap_16(x) OSSwapInt16(x)
 #define bswap_32(x) OSSwapInt32(x)
 #define bswap_64(x) OSSwapInt64(x)
 
-#else
+#elif defined(__linux__)
 
-#ifdef __linux__
 #define PROG_OS "linux"
-#else
-#define PROG_OS "unknown"
-#endif
-
 #include <sys/sysinfo.h>
 #include <byteswap.h>
+#include <sys/resource.h>
 
-#endif
+#elif defined(__FreeBSD__)
 
+#define PROG_OS "freebsd"
+#include <sys/sysinfo.h>
+#include <sys/resource.h>
+#include <sys/endian.h>
+#define bswap_16(x) bswap16(x)
+#define bswap_32(x) bswap32(x)
+#define bswap_64(x) bswap64(x)
+
+#elif defined(__NetBSD__)
+
+#define PROG_OS "netbsd"
+#include <sys/resource.h>
+#include <sys/types.h>
+#include <machine/bswap.h>
+#define bswap_16(x) bswap16(x)
+#define bswap_32(x) bswap32(x)
+#define bswap_64(x) bswap64(x)
+// Alters behavior, but NetBSD 7 does not have getopt_long_only()
+#define getopt_long_only getopt_long
+
+#else	// Other Unix
+
+#define PROG_OS "unknown"
+#include <sys/sysinfo.h>
+#include <byteswap.h>
 #include <sys/resource.h>
 
 #endif
```

**Comment by torognes** [comment]
Thanks!

I've made the changes to `city.h` and `vsearch.h` as suggested. I have also added a note in the manual due to the missing `getopt_long_only` on NetBSD (only double dash options accepted).

I have not made the changes to `configure.ac` and `Makefile.am` that are in the patches on freebsd.org. I understand that the way the `CFLAGS` and `CXXFLAGS` are modified in `configure.ac` and `Makefile.am` may not be according to the standard, because it overrides the default values and those specified on the command line with `configure`. The main reason for this is due to the use of a SSSE3 instruction (PSHUFB) in `cpu.cc`. The SSSE3 instructions may not be available on all x86_64 systems. VSEARCH will detect the CPU features at run-time and select the appropriate code. The file `cpu.cc` is compiled twice with and without SSSE3 enabled. I have not found a good way to achieve this without overriding those options. But perhaps it's time to let go with those old systems that do not have SSSE3.

**Comment by frederic-mahe** [comment]
> The main reason for this is due to the use of a SSSE3 instruction (PSHUFB) in cpu.cc. The SSSE3 instructions may not be available on all x86_64 systems. VSEARCH will detect the CPU features at run-time and select the appropriate code. The file cpu.cc is compiled twice with and without SSSE3 enabled. I have not found a good way to achieve this without overriding those options. But perhaps it's time to let go with those old systems that do not have SSSE3.

This could help you make a decision: [Most recent SSSE2-capable x86 processor without support of SSSE3 instructions](https://stackoverflow.com/questions/52858556/most-recent-processor-without-support-of-ssse3-instructions) (in short, 2012 for AMD and 2006 for Intel).

**Comment by outpaddling** [comment]
Cool, thanks.

I suspected there were good reasons for the hard-coded options in configure.av and Makefile.am, so I intended to leave them as patches in the FreeBSD port and pkgsrc package frameworks rather than push them upstream.

Your runtime code-selection system is an interesting way around the limitations of binary packages.

Not necessary for build-from source package managers like FreeBSD ports and pkgsrc, though.  Users can either use the binary package and accept the "portable" optimizations, or build any package with any additions flags they like.

Out of curiosity, have you tested whether vsearch benefits much from AVX?
